hack/e2e/cmd/generateServicePrincipals.go

```diff
@@ -0,0 +1,218 @@
+		}
+
+		log.Infof("found IDs for '%s': ObjectID='%s', AppID='%s'", spnName, spnObjectID, spnAppID)
+
+		log.Info("Ensuring SPN ID in keyvault...")
+
+		err = keyvaultManager.SetSecret(ctx, fmt.Sprintf("%s-app-id", spnName), azkeyvault.SecretSetParameters{
+			Value: to.StringPtr(spnAppID),
+		})
+		if err != nil {
+			return err
+		}
+
+		if len(apps[0].GetPasswordCredentials()) > 0 {
+			log.Infof("clearing %d passwords on SPN", len(apps[0].GetPasswordCredentials()))
+
+			for _, n := range apps[0].GetPasswordCredentials() {
+				body := msgraph_apps.NewItemRemovePasswordPostRequestBody()
+				body.SetKeyId(n.GetKeyId())
+
+				err = graphClient.Applications().ByApplicationId(spnObjectID).RemovePassword().Post(ctx, body, nil)
+				if err != nil {
+					return err
+				}
+			}
+		} else {
+			log.Info("no passwords to remove")
+		}
+
+		log.Info("creating password credential for SPN")
+
+		endDateTime := time.Now().UTC().AddDate(0, 0, 90)
+		expiryTime := date.NewUnixTimeFromSeconds(float64(endDateTime.Unix()))
+
+		pwCredential := msgraph_models.NewPasswordCredential()
+		pwCredential.SetDisplayName(&spnName)
+		pwCredential.SetEndDateTime(&endDateTime)
+
+		pwCredentialRequestBody := msgraph_apps.NewItemAddPasswordPostRequestBody()
+		pwCredentialRequestBody.SetPasswordCredential(pwCredential)
+
+		resp, err := graphClient.Applications().ByApplicationId(spnObjectID).AddPassword().Post(ctx, pwCredentialRequestBody, nil)
+		if err != nil {
+			return err
+		}
+
+		log.Info("setting password credential for SPN in keyvault")
+		err = keyvaultManager.SetSecret(ctx, fmt.Sprintf("%s-secret-value", spnName), azkeyvault.SecretSetParameters{
+			Value: resp.GetSecretText(),
+			SecretAttributes: &azkeyvault.SecretAttributes{
+				Expires: &expiryTime,
+			},
+		})
+		if err != nil {
+			return err
+		}
+
+		log.Info("checking for service principal")
+
+		filter = fmt.Sprintf("appId eq '%s'", spnAppID)
+		spResult, err := graphClient.ServicePrincipals().Get(ctx, &serviceprincipals.ServicePrincipalsRequestBuilderGetRequestConfiguration{
+			QueryParameters: &serviceprincipals.ServicePrincipalsRequestBuilderGetQueryParameters{
+				Filter: &filter,
+			},
+		})
+		if err != nil {
+			return err
+		}
+		sps := spResult.GetValue()
+
+		spID := ""
+
+		switch len(sps) {
+		case 0:
+			log.Info("found no existing service principals, creating")
+
+			requestBody := msgraph_models.NewServicePrincipal()
+			requestBody.SetAppId(&spnAppID)
+			spAddResult, err := graphClient.ServicePrincipals().Post(ctx, requestBody, nil)
+			if err != nil {
+				return err
+			}
+
+			spID = *spAddResult.GetId()
+		case 1:
+			spID = *sps[0].GetId()
+			log.Infof("found SP ID %s", spID)
+		default:
+			return fmt.Errorf("%d sps found for spn '%s'", len(sps), spnName)
+		}
+
+		log.Info("setting service principal ID in keyvault")
+		err = keyvaultManager.SetSecret(ctx, fmt.Sprintf("%s-sp-id", spnName), azkeyvault.SecretSetParameters{
```

Comment:
Same as before for _azkeyvault.SecretSetParameters_. Even here is not that important, I still see value in extracting it.